troubleshooting/checkpointing/src/solution/java/com/ververica/flink/training/solutions/CheckpointingJobSolution3.java

```diff
@@ -0,0 +1,300 @@
+		}
+
+		env.execute(CheckpointingJobSolution3.class.getSimpleName());
+	}
+
+	public static class SortMeasurementFunction
+			extends KeyedProcessFunction<Integer, Tuple2<Measurement, Long>, Tuple2<Measurement, Long>> {
+
+		private ListState<Tuple2<Measurement, Long>> listState;
+
+		@Override
+		public void open(Configuration parameters) throws Exception {
+			super.open(parameters);
+
+			ListStateDescriptor<Tuple2<Measurement, Long>> desc =
+					new ListStateDescriptor<Tuple2<Measurement, Long>>(
+							"events",
+							TypeInformation.of(new TypeHint<Tuple2<Measurement, Long>>(){})
+					);
+			listState = getRuntimeContext().getListState(desc);
+		}
+
+		@Override
+		public void processElement(Tuple2<Measurement, Long> value, Context ctx,
+								   Collector<Tuple2<Measurement, Long>> out) throws Exception {
+			TimerService timerService = ctx.timerService();
+
+			if (ctx.timestamp() > timerService.currentWatermark()) {
+				listState.add(value);
+				timerService.registerEventTimeTimer(ctx.timestamp());
+			}
+		}
+
+		@Override
+		public void onTimer(long timestamp, OnTimerContext ctx,
+							Collector<Tuple2<Measurement, Long>> out) throws Exception {
+
+			ArrayList<Tuple2<Measurement, Long>> list = new ArrayList<>();
+			listState.get().iterator().forEachRemaining(list::add);
+			list.sort(new MeasurementByTimeComparator());
+
+			Long watermark = ctx.timerService().currentWatermark();
+			int index = 0;
+			for (Tuple2<Measurement, Long> event : list) {
+				if (event != null && event.f1 <= watermark) {
+					out.collect(event);
+					index++;
+				} else {
+					break;
+				}
+			}
+			list.subList(0, index).clear();
+			listState.update(list);
+		}
+	}
+
+	public static class MeasurementWindowAggregatingFunction implements
+			AggregateFunction<Tuple2<Measurement, Long>, Tuple3<Long, Double, Double>, Tuple2<Long, Double>> {
+		private static final long serialVersionUID = 1;
+
+		@Override
+		public Tuple3<Long, Double, Double> createAccumulator() {
+			/**
+			 * f0: the total number of events
+			 * f1: the total differences summed up in the event time order
+			 * f2: the value of the previous measurement
+			 */
+			return new Tuple3<Long, Double, Double>(0L, 0.0, 0.0);
+		}
+
+		@Override
+		public Tuple3<Long, Double, Double> add(
+				final Tuple2<Measurement, Long> record,
+				final Tuple3<Long, Double, Double> aggregate) {
+
+			if (aggregate.f0 > 0) {
+				aggregate.f1 += record.f0.getValue() - aggregate.f2;
+			}
+			aggregate.f0++;
+			aggregate.f2 = record.f0.getValue();
+			return aggregate;
+		}
+
+		@Override
+		public Tuple2<Long, Double> getResult(final Tuple3<Long, Double, Double> windowedMeasurements) {
+			return new Tuple2<>(windowedMeasurements.f0, windowedMeasurements.f1);
+		}
+
+		@Override
+		public Tuple3<Long, Double, Double> merge(
+				final Tuple3<Long, Double, Double> agg1,
+				final Tuple3<Long, Double, Double> agg2) {
+			// not needed in this case
+			return null;
```

Comment:
Rather throw an `UnsupportOperationException` to fail early (because this way of aggregating doesn't even support merging two windows since then the results are wrong)?

Reply:
Right, throwing an unsupported exception is a good idea. On the other hand, do you know in which situation this method is called? For this particular exercise job, I do not think of any.